publisher/sonatype-nx3/src/main/java/eu/maveniverse/maven/njord/publisher/sonatype/nx3/SonatypeNx3Publisher.java

```diff
@@ -0,0 +1,344 @@
+                                + "Tagging is only available in Nexus Repository Pro. The tag will be omitted.",
+                        publisherConfig.tag());
+            }
+
+            for (Map.Entry<String, List<Artifact>> entry : componentGroups.entrySet()) {
+                String gav = entry.getKey();
+                List<Artifact> artifacts = entry.getValue();
+                uploadComponent(
+                        httpClient,
+                        repository,
+                        authSource,
+                        repositoryName,
+                        extraHeaders,
+                        gav,
+                        artifacts,
+                        artifactStore,
+                        isProEdition);
+            }
+        }
+
+        logger.info("Successfully published {} component(s)", componentGroups.size());
+    }
+
+    /**
+     * Groups artifacts by their GAV coordinates (groupId:artifactId:version).
+     * Filters out metadata files and checksum files.
+     */
+    private Map<String, List<Artifact>> groupArtifactsByGav(ArtifactStore artifactStore) {
+        Map<String, List<Artifact>> groups = new LinkedHashMap<>();
+
+        for (Artifact artifact : artifactStore.artifacts()) {
+            // Skip metadata files
+            if (artifact.getExtension().equals("xml")
+                    && artifact.getArtifactId().equals("maven-metadata")) {
+                continue;
+            }
+
+            // Skip checksum files (they're auto-generated by NXRM3)
+            if (artifact.getExtension().equals("md5")
+                    || artifact.getExtension().equals("sha1")
+                    || artifact.getExtension().equals("sha256")
+                    || artifact.getExtension().equals("sha512")) {
+                continue;
+            }
+
+            String gav = artifact.getGroupId() + ":" + artifact.getArtifactId() + ":" + artifact.getVersion();
+            groups.computeIfAbsent(gav, k -> new ArrayList<>()).add(artifact);
+        }
+
+        return groups;
+    }
+
+    /**
+     * Uploads a single component (GAV group) to NXRM3 using the Components API.
+     */
+    private void uploadComponent(
+            CloseableHttpClient httpClient,
+            RemoteRepository repository,
+            RemoteRepository authSource,
+            String repositoryName,
+            Map<String, String> extraHeaders,
+            String gav,
+            List<Artifact> artifacts,
+            ArtifactStore artifactStore,
+            boolean isProEdition)
+            throws IOException {
+        try {
+            URIBuilder uriBuilder = new URIBuilder(repository.getUrl());
+            uriBuilder.setPath("/service/rest/v1/components");
+            uriBuilder.addParameter("repository", repositoryName);
+
+            HttpPost post = new HttpPost(uriBuilder.build());
+            extraHeaders.forEach(post::setHeader);
+
+            // Add authentication if available
```

Comment:
I would rather rely on the [BasicCredentialsProvider](https://hc.apache.org/httpcomponents-client-4.5.x/current/httpclient/apidocs/org/apache/http/impl/client/BasicCredentialsProvider.html) to set the authentication headers.